source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -0,0 +1,209 @@
+#include "extensions/filters/http/cache/cache_filter.h"
+
+#include "envoy/registry/registry.h"
+
+#include "common/config/utility.h"
+#include "common/http/headers.h"
+
+#include "absl/memory/memory.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+
+bool CacheFilter::isCacheableRequest(Http::HeaderMap& headers) {
+  const Http::HeaderEntry* method = headers.Method();
+  const Http::HeaderEntry* forwarded_proto = headers.ForwardedProto();
+  const Http::HeaderValues& header_values = Http::Headers::get();
+  // TODO(toddmgreer): Also serve HEAD requests from cache.
+  // TODO(toddmgreer): Check all the other cache-related headers.
+  return method && forwarded_proto && headers.Path() && headers.Host() &&
+         (method->value() == header_values.MethodValues.Get) &&
+         (forwarded_proto->value() == header_values.SchemeValues.Http ||
+          forwarded_proto->value() == header_values.SchemeValues.Https);
+}
+
+bool CacheFilter::isCacheableResponse(Http::HeaderMap& headers) {
+  const Http::HeaderEntry* cache_control = headers.CacheControl();
+  // TODO(toddmgreer): fully check for cacheability. See for example
+  // https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/http/caching_headers.h.
+  if (cache_control) {
+    return !StringUtil::caseFindToken(cache_control->value().getStringView(), ",",
+                                      Http::Headers::get().CacheControlValues.Private);
+  }
+  return false;
+}
+
+HttpCache&
+CacheFilter::getCache(const envoy::config::filter::http::cache::v2::CacheConfig& config) {
+  return Config::Utility::getAndCheckFactory<HttpCacheFactory>(config).getCache(config);
+}
+
+CacheFilter::CacheFilter(const envoy::config::filter::http::cache::v2::CacheConfig& config,
+                         const std::string&, Stats::Scope&, TimeSource& time_source)
+    : time_source_(time_source), cache_(getCache(config)) {}
+
+void CacheFilter::onDestroy() {
+  lookup_ = nullptr;
+  insert_ = nullptr;
+}
```

Comment:
Can you add a small comment here that we assume that destroying the handles cancels all async callbacks, etc.?